src/ActiveLogin.Authentication.BankId.AspNetCore/Areas/BankIdAuthentication/Controllers/BankIdController.cs

```diff
@@ -21,34 +23,55 @@ public class BankIdController : Controller
         private readonly IBankIdUserMessageLocalizer _bankIdUserMessageLocalizer;
         private readonly IBankIdLoginOptionsProtector _loginOptionsProtector;
         private readonly IStringLocalizer<BankIdHandler> _localizer;
+        private readonly IBankIdInvalidStateHandler _bankIdInvalidStateHandler;
 
         public BankIdController(
             IAntiforgery antiforgery,
             IBankIdUserMessageLocalizer bankIdUserMessageLocalizer,
             IBankIdLoginOptionsProtector loginOptionsProtector,
-            IStringLocalizer<BankIdHandler> localizer)
+            IStringLocalizer<BankIdHandler> localizer,
+            IBankIdInvalidStateHandler bankIdInvalidStateHandler)
         {
             _antiforgery = antiforgery;
             _bankIdUserMessageLocalizer = bankIdUserMessageLocalizer;
             _loginOptionsProtector = loginOptionsProtector;
             _localizer = localizer;
+            _bankIdInvalidStateHandler = bankIdInvalidStateHandler;
         }
 
         [HttpGet]
-        public ActionResult Login(string returnUrl, string loginOptions)
+        public async Task<ActionResult> Login(string returnUrl, string loginOptions)
         {
             if (!Url.IsLocalUrl(returnUrl))
             {
                 throw new Exception(BankIdConstants.InvalidReturnUrlErrorMessage);
             }
 
             var unprotectedLoginOptions = _loginOptionsProtector.Unprotect(loginOptions);
-            var antiforgeryTokens = _antiforgery.GetAndStoreTokens(HttpContext);
+            if (!HasStateCookie(unprotectedLoginOptions))
+            {
+                var invalidStateContext = new BankIdInvalidStateContext(unprotectedLoginOptions.CancelReturnUrl);
```

Comment:
CHECK: Any reason we do not want to pass in all the options here?

Reply:
Yes - we gave it a thought and only want to expose as little as possible. The very moment we expose a thing, we need to maintain it for some time. In this case I only want to provide data that I think makes sense, which in this case is CancelReturnUrl. If we have a usecase for another of tjose properties I'm willing to add that to the "context" as well :)